Google.Api.Gax.Grpc/RetryAttempt.cs

```diff
@@ -0,0 +1,121 @@
+
+        /// <summary>
+        /// Returns a sequence of retry attempts. The sequence has <see cref="RetrySettings.MaxAttempts"/> elements, and calling
+        /// <see cref="ShouldRetry(Exception)"/> on the last attempt will always return false. This overload assumes no deadline,
+        /// and so does not require a clock.
+        /// </summary>
+        /// <param name="settings">The retry settings to create a sequence for. Must not be null.</param>
+        /// <param name="scheduler">The scheduler to use for delays.</param>
+        /// <param name="initialBackoffOverride">An override value to allow an initial backoff which is not the same
+        /// as <see cref="RetrySettings.InitialBackoff"/>. This is typically to allow an "immediate first retry".</param>
+        /// <returns></returns>
+        public static IEnumerable<RetryAttempt> CreateRetrySequence(RetrySettings settings, IScheduler scheduler, TimeSpan? initialBackoffOverride = null)
+        {
+            GaxPreconditions.CheckNotNull(settings, nameof(settings));
+            GaxPreconditions.CheckNotNull(scheduler, nameof(scheduler));
+
+            TimeSpan backoff = initialBackoffOverride ?? settings.InitialBackoff;
+            for (int i = 1; i <= settings.MaxAttempts; i++)
+            {
+                yield return new RetryAttempt(settings, null, null, scheduler, i, settings.BackoffJitter.GetDelay(backoff));
+                backoff = settings.NextBackoff(backoff);
+            }
+        }
+
+        /// <summary>
+        /// Returns a sequence of retry attempts. The sequence has <see cref="RetrySettings.MaxAttempts"/> elements, and calling
+        /// <see cref="ShouldRetry(Exception)"/> on the last attempt will always return false.
+        /// </summary>
+        /// <param name="settings">The retry settings to create a sequence for. Must not be null.</param>
+        /// <param name="scheduler">The scheduler to use for delays.</param>
+        /// <param name="deadline">The overall deadline for the operation.</param>
+        /// <param name="clock">The clock to use to compare the current time with the deadline.</param>
+        /// <param name="initialBackoffOverride">An override value to allow an initial backoff which is not the same
+        /// as <see cref="RetrySettings.InitialBackoff"/>. This is typically to allow an "immediate first retry".</param>
+        /// <returns></returns>
+        public static IEnumerable<RetryAttempt> CreateRetrySequence(RetrySettings settings, IScheduler scheduler, DateTime? deadline, IClock clock, TimeSpan? initialBackoffOverride = null)
+        {
+            GaxPreconditions.CheckNotNull(settings, nameof(settings));
+            GaxPreconditions.CheckNotNull(clock, nameof(clock));
+            GaxPreconditions.CheckNotNull(scheduler, nameof(scheduler));
+
+            // It's simpler not to need nullable logic when computing deadlines.
+            var effectiveDeadline = deadline ?? DateTime.MaxValue;
```

Comment:
Or maybe, since there is the other overload now, make the deadline non-nullable in this one?

Reply:
No, as there are cases where we might or might not have a deadline - quite a bit of code uses `DateTime?` for an effective deadline. It would be nasty to have to conditionalise which overload is called.